src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,181 @@
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is
+-- the length of the given list.
+--
+-- @
+-- vertices []        == 'empty'
+-- vertices [x]       == 'vertex' x
+-- vertices [1, 2, 3] == 1 + 2 + 3
+-- @
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices xs = AAM $ AM.vertices xs
+
+-- | Perform a disjoint overlay of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- overlayD empty empty             == 'empty'
+-- overlayD (1 * 2 + 1 * 3) (1 * 2) == (Left 1) * (Left 2) 
+--                                   + (Left 1) * (Left 3)
+--                                   + (Right 1) * (Right 2)
+-- @
+overlayD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+overlayD (AAM a) (AAM b) = AAM (AM.overlay (AM.gmap Left a) (AM.gmap Right b))
+
+-- | Perform a disjoint connect of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- connectD empty empty     == 'empty'
+-- connectD (1 + 2) (1 + 2) == (Left 1) * (Right 1) 
+--                           + (Left 1) * (Right 2)
+--                           + (Left 2) * (Right 1)
+--                           + (Left 2) * (Right 2)
+-- connectD (1 * 2) (1 * 2) == (Left 1) * (Right 1)
+--                           + (Left 1) * (Right 2) 
+--                           + (Left 2) * (Right 1) 
+--                           + (Left 2) * (Right 2) 
+--                           + (Left 1) * (Left 2) 
+--                           + (Right 1) * (Right 2)
+-- @
+connectD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+connectD (AAM a) (AAM b) = AAM (AM.connect (AM.gmap Left a) (AM.gmap Right b))
+
+-- | Compute the /transitive closure/ of a graph.
+-- Complexity: /O(n * m * log(n)^2)/ time.
+--
+-- @
+-- transitiveClosure 'empty'             == 'empty'
+-- transitiveClosure ('vertex' x)        == 'vertex' x
+-- transitiveClosure (1 * 2 + 2 * 3)     == 1 * 2 + 2 * 3 + 1 * 3
+-- transitiveClosure . transitiveClosure == transitiveClosure
+-- @
+transitiveClosure :: Ord a => AdjacencyMap a -> AdjacencyMap a
+transitiveClosure (AAM x) = AAM (AM.transitiveClosure x)
+
+-- | Compute the /condensation/ of a graph, where each vertex
+-- corresponds to a /strongly-connected component/ of the original
+-- graph. Note that component graphs are non-empty, and are therefore
+-- of type "Algebra.Graph.NonEmpty.AdjacencyMap".
+--
+-- @
+-- scc 'empty'      == 'empty'
+-- scc ('vertex' x) == 'vertex' (NonEmpty.'NonEmpty.vertex' x)
+-- @
+scc :: (Ord a) => AM.AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
+scc = AAM . AM.scc
+
+-- | Compute the /topological sort/ of a graph.
+-- 
+-- @
+-- topSort (1)             == [1]
+-- topSort (1 * 2 + 3 * 1) == [3,1,2]
+-- @
+topSort :: (Ord a) => AdjacencyMap a -> [a]
+topSort (AAM am) = Typed.topSort (Typed.fromAdjacencyMap am)
```

Comment:
Why are you going via the `Typed` module? I think you should just do `fromJust . AM.topSort`.

Reply:
Sorry, my above comment is wrong. Let's keep your current implementation, because going via `AM.topSort` leads to unnecessary `isTopSort` check.